raiden/transfer/log.py

```diff
@@ -0,0 +1,238 @@
+            'SELECT * from transactions'
+        )
+        return result.fetchall()
+
+    def read(self):
+        pass
+
+    def last_identifier(self):
+        cursor = self.conn.cursor()
+        result = cursor.execute(
+            'SELECT seq FROM sqlite_sequence WHERE name="transactions"'
+        )
+        result = result.fetchall()
+        if result != list():
+            assert len(result) == 1
+            result = result[0][0]
+        else:
+            result = 0
+        return result
+
+    def __del__(self):
+        self.conn.close()
+
+
+class TransactionLogFileBackend(TransactionLogStorageBackend):
+    """This is just an example for having a file backend. Not actually implemented"""
+
+    def __init__(self, filepath):
+        self.filepath = filepath
+        self.file = create_file_iff_not_existing(
+            self.filepath,
+            return_it=True,
+            mode='r+',
+            buffering=False
+        )
+        self._synced = False  #: True when the existing log is read to the end
+
+    def write_transaction(self, data):
+        pass
+
+    def write_state_snapshot(self, identifier, data):
+        pass
+
+    def read(self):
+        pass
+
+    def last_identifier(self):
+        pass
+
+    def __del__(self):
+        self.file.flush()
+        self.file.close()
+
+
+class TransactionLog(object):
+
+    def __init__(
+            self,
+            storage_class,
+            serializer_class=PickleTransactionSerializer()):
+
+        if not issubclass(type(serializer_class), TransactionLogSerializer):
+            raise ValueError('serializer_class must follow the TransactionLogSerializer interface')
+        self.serializer = serializer_class
+
+        if not issubclass(type(storage_class), TransactionLogStorageBackend):
+            raise ValueError(
+                'storage_class must follow the TransactionLogStorageBackend interface'
+            )
+        self.storage = storage_class
+
+    def log(self, state_change):
+        # TODO: Issue 587
+        # Implement a queue of state changes for batch writting
+        serialized_data = self.serializer.serialize(state_change)
+        self.storage.write_transaction(serialized_data)
+
+    def get_transaction_by_id(self, identifier):
+        serialized_data = self.storage.get_transaction_by_id(identifier)
+        return self.serializer.deserialize(serialized_data)
+
+    def last_identifier(self):
+        return self.storage.last_identifier()
+
+    def get_all_state_changes(self):
+        """ Returns a list of tuples of identifiers and state changes"""
+        return [
+            (res[0], self.serializer.deserialize(res[1]))
+            for res in self.storage.get_all_transactions()
+        ]
+
+    def snapshot(self, state):
+        serialized_data = self.serializer.serialize(state)
+        self.storage.write_state_snapshot(self.last_identifier(), serialized_data)
+
+
+def unapplied_state_changes(transaction_log, snapshot):
```

Comment:
This function makes sense for the file backend, but with the sqllite backend we can filter the unapplied changes with SQL, so IMO this should be removed if `TransactionLogFileBackend` is removed.